### PR TITLE
add postgres db client for cockroachdb

### PR DESCRIPTION
**Pull request by jgd5**
Ran  ./node_modules/.bin/ts-node -T test/engine/index.js with a local cockraochdb.

A couple remaining issues requiring your inputs:

* What is the best way to configure which database implementation to use?
* How do we plan to run tests?   Download the cockroach binary (it's about 200MB) on the fly or preconfigure cockroachdb on test machine.

### CODE REVIEW

File: lib/engine/db/postgres/schema.sql [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+--    cockroach start-single-node  --insecure --listen-addr=localhost:8888 --store=type=mem,size=640MiB
+--     
+-- The following command can be use to initialize the databse for testing:
+--    cockroach sql --url postgresql://root@localhost:8888?sslmode=disable < schema.sql 
+
+drop database if exists thingengine;
+create database thingengine; 
+use thingengine;
+
+drop table if exists app ;
```
Comment (gcampax) [inline lib/engine/db/postgres/schema.sql:15]: The only tables we need are `app`, `channel`, `device`, `device_journal`. The rest can go.

File: lib/engine/db/postgres/index.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+                reject(err);
+            else
+                resolve();
+        });
+    });
+}
+
+export function selectAll(client : PoolClient, string : string, args : unknown[]) {
+    return new Promise<any[]>((resolve, reject) => {
+        client.query(string, args, (err, res) => {
```
Comment (gcampax) [inline lib/engine/db/postgres/index.ts:48]: I swear I originally modeled the `query`, `selectAll`, `selectOne` APIs based on postgres (many years ago before deciding on mysql for almond-cloud instead). Maybe that was a different postgres library?
Anyway interesting that we end up wrapping anyway.

File: lib/engine/db/postgres/index.ts [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                });
+                this._transactionQueue.set(client, newqueue);
+            });
+        });
+    }
+
+    ensureSchema() {
+        // Database is initialized out of band
+        // eslint-disable-next-line no-promise-executor-return
+        return new Promise<void>((resolve, reject) => resolve());
```
Comment (gcampax) [inline lib/engine/db/postgres/index.ts:140]:
```suggestion
        return Promise.resolve();
```
(alternatively, make the `ensureSchema()` function async and leave it empty)

File: lib/engine/db/postgres/index.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+                }
+                resolve(callback(c));
+                release();
+            });
+        });
+    }
+
+    withTransaction<T>(transaction : (client : PoolClient) => Promise<T>) : Promise<T> {
+        return this.withClient((client) => {
+            const queue = this._transactionQueue.get(client) || Promise.resolve();
```
Comment (gcampax) [inline lib/engine/db/postgres/index.ts:115]: We definitely don't need the transaction queue here, this was a hack to fix broken sqlite's locking. With a real DB, we have real locking and can run multiple transactions concurrently...

**Review by gcampax — COMMENTED** [review]
This looks ok, but i wonder what's the best approach for modularity: should the whole postgresdb implementation live in almond-cloud (where we can deal with schema changes centrally)? Then you'd call [`platform.getCapability`](https://almond-static.stanford.edu/jsdoc/thingpedia/latest/classes/baseplatform.html#getcapability) and get back the DB interface directly.
Partially also because I want to keep the dependencies of genie-toolkit somewhat lean, given it goes into all sorts of docker images.

Alternatively, a generic `platform.getDBConfig()` that returns an object, or a `platform.getDatabaseUrl()`, and we use the scheme to distinguish the backend.

Also some comments on the implementation.

**Comment by jgd5** [comment]
Abandon this PR in favor of #648

**Pull request closed without merging**